experimental/iree_profiler/matmul.py

```diff
@@ -0,0 +1,580 @@
+    self.batch_size = args.batch_size
+    self.benchmark_repetitions = args.benchmark_repetitions
+    self.verbose = False if args.verbose in ['False', 'false', '0'] else True
+
+    # Additional paths.
+    self.matmul_path = os.path.join(self.generated_path, 'matmul')
+    self.operation_path = os.path.join(self.matmul_path, operation.name())
+    self.source_mlir_file = os.path.join(self.operation_path,
+                                         operation.name() + '.mlir')
+
+    # path to iree-compile tool. (for compiling the input mlir file to vmfb)
+    self.iree_compile_path = os.path.join(args.build_dir, 'tools/iree-compile')
+    self.force_compile = False if args.force_compile in ['False', 'false', '0'
+                                                        ] else True
+
+    # path to iree-benchmark-module tool. (for performance benchmarking and profiling)
+    self.iree_benchmark_module_path = os.path.join(
+        args.build_dir, 'tools/iree-benchmark-module')
+
+    # path to iree-run-module tool. (for verification)
+    self.iree_run_module_path = os.path.join(args.build_dir,
+                                             'tools/iree-run-module')
+
+    # output vmfb files for the operation.
+    self.vmfb_verify_file = os.path.join(self.operation_path,
+                                         self.operation.name() + '_verify.vmfb')
+    self.vmfb_benchmark_file = os.path.join(
+        self.operation_path,
+        self.operation.name() + '_benchmark.vmfb')
+
+  # Compile the matmul operation to a vmfb file for benchmarking.
+  def compile(self, compilation_mode):
+
+    cmd_template = \
+    "${iree_compile_path} ${source_mlir_file} --iree-hal-target-backends=${device} "\
+    "--iree-hal-cuda-llvm-target-arch=sm_80 --iree-hal-benchmark-dispatch-repeat-count=${benchmark-dispatch-repeat-count} "\
+    "--iree-codegen-llvmgpu-use-mma-sync=false -o ${vmfb_file}"
+
+    benchmark_dispatch_repeat_count = self.benchmark_dispatch_repeat_count if compilation_mode == CompilationMode.Benchmark else 1
+    vmfb_file = self.vmfb_benchmark_file if compilation_mode == CompilationMode.Benchmark else self.vmfb_verify_file
+
+    values = {
+        'iree_compile_path': self.iree_compile_path,
+        'source_mlir_file': self.source_mlir_file,
+        'device': self.device,
+        'benchmark-dispatch-repeat-count': str(benchmark_dispatch_repeat_count),
+        'vmfb_file': vmfb_file
+    }
+
+    if not os.path.exists(vmfb_file) or self.force_compile:
+      cmd = SubstituteTemplate(cmd_template, values)
+      print(">> Compilation command for " +
+            CompilationModeNames[compilation_mode] + " : " + cmd)
+      subprocess.getoutput(cmd)
+
+    else:
+      if self.verbose:
+        print("Skipping compilation of matmul operation: " + vmfb_file +
+              " since it already exists.")
+
+  # Verify a matmul operation with a given configuration.
+  def verify(self, configuration):
```

Comment:
Can much of this can be extracted to have a general utility function for verification using a list of input npys, a module, a function name, and an output npy? Similar for benchmarking.

Reply:
Yup, that would make the cmd line generation for verification and profiling + the launch call for the dispatch agnostic of the operation it is running. I am made a note of this in the document (which I will polish more) and we can take up this and other improvements and enhancements from the list.